content/articles/authentication-using-facebook-in-android-studio/index.md

```diff
@@ -0,0 +1,303 @@
+
+</androidx.constraintlayout.widget.ConstraintLayout>
+```
+
+### Step 9: The Code
+We will now write our code in the MainActivity
+
+```Kotlin
+import android.content.Intent
+import android.os.Bundle
+import android.view.View
+import android.widget.Toast
+import androidx.appcompat.app.AppCompatActivity
+import com.bumptech.glide.Glide
+import com.facebook.*
+import com.facebook.login.LoginResult
+import com.thecalvary.facebooklogindemo.databinding.ActivityMainBinding
+import org.json.JSONObject
+import java.util.*
+
+private const val TAG = "MainActivity"
+class MainActivity : AppCompatActivity() {
+
+    private lateinit var binding: ActivityMainBinding
+    private lateinit var callBackManager: CallbackManager
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        binding = ActivityMainBinding.inflate(layoutInflater)
+        val view: View = binding.root
+        setContentView(view)
+        callBackManager = CallbackManager.Factory.create()
+        binding.loginButton.setReadPermissions(listOf("email","public_profile","user_gender","user_birthday"))
+        binding.loginButton.registerCallback(callBackManager, object : FacebookCallback<LoginResult>{
+            override fun onCancel() {
+                Toast.makeText(this@MainActivity,
+                    "Cancelled",
+                    Toast.LENGTH_SHORT).show() }
+
+            override fun onError(error: FacebookException) {
+                Toast.makeText(this@MainActivity, "$error", Toast.LENGTH_SHORT).show() }
+
+            override fun onSuccess(result: LoginResult) {
+                val graphRequest = GraphRequest.newMeRequest(result?.accessToken){`object` ,response ->
+                getFacebookData(`object`)
+                }
+                val parameters = Bundle()
+                parameters.putString("fields", "id,email,birthday,gender,name")
+                graphRequest.parameters = parameters
+                graphRequest.executeAsync()
+            } })
+    }
+
+    private fun getFacebookData(jsonObject: JSONObject?) {
+        val profilePic = "https://graph.facebook.com/${jsonObject
+?.getString("id")}/picture?width=500&height=500"
+        Glide.with(this)
+            .load(profilePic)
+            .into(binding.profilePic)
+
+        val name = jsonObject?.getString("name")
+        val birthday = jsonObject?.getString("birthday")
+        val gender = jsonObject?.getString("gender")
+        val email = jsonObject?.getString("email")
+
+        binding.userName.text = "Name: ${name}"
+        binding.userEmail.text = "Email: ${email}"
+        binding.userBDay.text = "Birthday: ${birthday}"
+        binding.userGender.text = "Gender: ${gender}"
+}
+     override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        super.onActivityResult(requestCode, resultCode, data)
+        callBackManager.onActivityResult(requestCode, resultCode, data) }
+}
+```
+
```

Comment:
Explain your code rather than just pasting it in the article

Reply:
Okay sure. Let me add the explanation

Reply:
I have added the comments as asked